### PR TITLE
feat: added overflow flags to trunc in InstCombine

### LINKED ISSUE #793: llvm: add [nsw|nuw] support to `trunc` operation
The following test case fails due to lack of nsw/nuw support:

```lean
theorem trunc_nuw_xor_thm (e e_1 : IntW 2) :
  trunc 1 (LLVM.xor e_1 e) ⊑ icmp IntPredicate.ne e_1 e := by 
    simp_alive_undef
    simp_alive_ops
    simp_alive_case_bash
    simp_alive_split
    simp_alive_benchmark
    all_goals sorry
```

### PR DESCRIPTION
**Pull request by lfrenot**

### CODE REVIEW

File: SSA/Projects/InstCombine/Base.lean [hunk 1]
```diff
@@ -174,6 +173,10 @@ namespace MOp
     := .binary w (.or DisjointFlag )
 
 /- These definitions use NoWrapFlags -/
+@[match_pattern] def trunc  (w w' : Width φ)
+  (NoWrapFlags: NoWrapFlags := {nsw := false , nuw := false}) : MOp φ
```
Comment (bollu) [inline SSA/Projects/InstCombine/Base.lean]:
```suggestion
@[match_pattern] def trunc  (w w' : Width φ)
  (noWrapFlags : NoWrapFlags := {nsw := false , nuw := false}) : MOp φ
```

File: SSA/Projects/InstCombine/Base.lean [hunk 2]
```diff
@@ -205,7 +208,7 @@ namespace MOp
 def deepCasesOn {motive : ∀ {φ}, MOp φ → Sort*}
     (neg    : ∀ {φ} {w : Width φ},              motive (neg  w))
     (not    : ∀ {φ} {w : Width φ},              motive (not  w))
-    (trunc  : ∀ {φ} {w w' : Width φ},              motive (trunc w w'))
+    (trunc  : ∀ {φ NoWrapFlags} {w w' : Width φ},              motive (trunc w w' NoWrapFlags))
```
Comment (bollu) [inline SSA/Projects/InstCombine/Base.lean]:
```suggestion
    (trunc  : ∀ {φ noWrapFlags} {w w' : Width φ},              motive (trunc w w' noWrapFlags))
```

File: SSA/Projects/InstCombine/Base.lean [hunk 3]
```diff
@@ -310,6 +312,9 @@ namespace Op
 @[match_pattern] abbrev or (w : Nat) (flag : DisjointFlag := {disjoint := false} ) : Op := MOp.or (.concrete w) flag
 
 /- These operations are separate from the others because they take in 2 flags: nuw and nsw.-/
+@[match_pattern] abbrev trunc (w w': Nat) (flags: NoWrapFlags :=
+   {nsw := false , nuw := false}) : Op := MOp.trunc (.concrete w) (.concrete w') flags
```
Comment (bollu) [inline SSA/Projects/InstCombine/Base.lean]:
```suggestion
@[match_pattern] abbrev trunc (w w': Nat) (noWrapFlags : NoWrapFlags :=
   {nsw := false , nuw := false}) : Op := MOp.trunc (.concrete w) (.concrete w') noWrapFlags
```

File: SSA/Projects/InstCombine/LLVM/EDSL.lean [hunk 4]
```diff
@@ -189,9 +189,19 @@ def mkExpr (Γ : Ctxt (MetaLLVM φ).Ty) (opStx : MLIR.AST.Op φ) :
...
+          let attr? := opStx.attrs.getAttr "overflowFlags"
+          match attr? with
+            | .none =>  pure <| .trunc (← getOutputWidth opStx "trunc")
+            | .some y => match y with
+              | .opaque_ "llvm.overflow" "nsw" => pure <| .trunc (← getOutputWidth opStx "trunc") ⟨true, false⟩
+              | .opaque_ "llvm.overflow" "nuw" => pure <| .trunc (← getOutputWidth opStx "trunc") ⟨false, true⟩
+              | .list [.opaque_ "llvm.overflow" "nuw", .opaque_ "llvm.overflow" "nsw"] => pure <| .trunc (← getOutputWidth opStx "trunc") ⟨true, true⟩
+              | .list [.opaque_ "llvm.overflow" "nsw", .opaque_ "llvm.overflow" "nuw"] => pure <| .trunc (← getOutputWidth opStx "trunc") ⟨true, true⟩
+              | .opaque_ "llvm.overflow" s => throw <| .generic s!"The overflow flag {s} not allowed. We currently support nsw (no signed wrap) and nuw (no unsigned wrap)"
+              | _ => throw <| .generic s!"Unrecognised overflow flag found: {MLIR.AST.docAttrVal y}. We currently support nsw (no signed wrap) and nuw (no unsigned wrap)"
```
Comment (bollu) [inline SSA/Projects/InstCombine/LLVM/EDSL.lean:206]: This does not make me happy, but sure, I guess this works. I would have assumed that the `nsw` and `nuw` are encoded as singleton lists, not as a top-level flag! Can you double-check that this is how MLIR does it? It seems like a very irregular encoding where the singleton version where we have only `nsw` (or `nuw`) is encoded differently from `[nsw, nuw]`.

File: SSA/Projects/InstCombine/LLVM/PrettyEDSLOverflow.lean [hunk 5]
```diff
@@ -36,4 +37,16 @@ macro_rules
       | _ => Macro.throwUnsupported
     `([mlir_op| $[$resName =]? $opName ($xs,*) $[<{overflowFlags = #llvm.overflow< $ovflags,* >}>]? : ($argTys,*) -> ($retTy:mlir_type,*) ])
 
+syntax mlir_op_operand " = " MLIR.Pretty.overflow_int_cast_op mlir_op_operand (" overflow<" overflow_flag,* "> ")? " : " mlir_type " to " mlir_type : mlir_op
+macro_rules
+  | `(mlir_op| $resName:mlir_op_operand = $name:MLIR.Pretty.overflow_int_cast_op $x $[overflow< $[$f],* >]? : $t to $t' ) => do
+    let some opName := extractOpName name.raw | Macro.throwUnsupported
+    let ovflags ← f.mapM fun f => f.mapM fun
```
Comment (bollu) [inline SSA/Projects/InstCombine/LLVM/PrettyEDSLOverflow.lean]: I don't know what `f` is supposed to be, but I would appreciate clearer variable names, please :)

File: SSA/Projects/InstCombine/Base.lean [hunk 6]
```diff
@@ -84,7 +84,7 @@ inductive MOp.UnaryOp (φ : Nat) : Type
   | neg
   | not
   | copy
-  | trunc (w' : Width φ)
+  | trunc (w' : Width φ) (nswnuw : NoWrapFlags := {nsw := false, nuw := false} )
```
Comment (bollu) [inline SSA/Projects/InstCombine/Base.lean]:
```suggestion
  | trunc (w' : Width φ) (noWrapFlags : NoWrapFlags := {nsw := false, nuw := false} )
```

File: SSA/Projects/InstCombine/LLVM/Semantics.lean [hunk 7]
```diff
@@ -571,9 +571,14 @@ def trunc? {w: Nat} (w': Nat) (x: BitVec w) : IntW w' := do
   some <| (BitVec.truncate w' x)
 
 @[simp_llvm_option]
-def trunc {w: Nat} (w': Nat) (x: IntW w) : IntW w' := do
+def trunc {w: Nat} (w': Nat) (x: IntW w) (flags : NoWrapFlags := {nsw := false , nuw := false}) : IntW w' := do
```
Comment (bollu) [inline SSA/Projects/InstCombine/LLVM/Semantics.lean]:
```suggestion
def trunc {w: Nat} (w': Nat) (x: IntW w) (noWrapFlags : NoWrapFlags := {nsw := false , nuw := false}) : IntW w' := do
```

**Review by bollu — CHANGES_REQUESTED** [review]
I like the overall PR very much! I did request some changes, for consistency and clarity :)

File: SSA/Projects/InstCombine/LLVM/EDSL.lean (hunk 4 above)
Comment (lfrenot) [reply SSA/Projects/InstCombine/LLVM/EDSL.lean:206]: In MLIR they are encoded with the following tablegen code
```
def IOFnone : I32BitEnumAttrCaseNone<"none">;
def IOFnsw  : I32BitEnumAttrCaseBit<"nsw", 0>;
def IOFnuw  : I32BitEnumAttrCaseBit<"nuw", 1>;

def IntegerOverflowFlags : I32BitEnumAttr<
    "IntegerOverflowFlags",
    "LLVM integer overflow flags",
    [IOFnone, IOFnsw, IOFnuw]> {
  let separator = ", ";
  let cppNamespace = "::mlir::LLVM";
  let genSpecializedAttr = 0;
  let printBitEnumPrimaryGroups = 1;
}

def LLVM_IntegerOverflowFlagsAttr :
    EnumAttr<LLVM_Dialect, IntegerOverflowFlags, "overflow"> {
  let assemblyFormat = "`<` $value `>`";
}
```
I don't know if that can be translated to lean-mlir
If you have a cleaner solution, I will gladly use it

**Comment by lfrenot** [comment]
This should address everything except the issue in EDSL.lean

**Comment by lfrenot** [comment]
Force pushed to fix the merge issues
@tobiasgrosser

**Comment by tobiasgrosser** [comment]
Let's merge this for now. @lfrenot, please address the remaining changes tomorrow.

**Pull request merged**